packages/block-editor/src/components/off-canvas-editor/block-edit-button.js

```diff
@@ -2,26 +2,208 @@
  * WordPress dependencies
  */
 import { edit } from '@wordpress/icons';
-import { Button } from '@wordpress/components';
-import { useDispatch } from '@wordpress/data';
-import { forwardRef } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+import { forwardRef, useMemo, useState } from '@wordpress/element';
+import { Button, Modal } from '@wordpress/components';
+import { useSelect, useDispatch } from '@wordpress/data';
+import { createBlock as create } from '@wordpress/blocks';
 
 /**
  * Internal dependencies
  */
 import { store as blockEditorStore } from '../../store';
 
+// copied from packages/block-library/src/page-list/edit.js
+
+// We only show the edit option when page count is <= MAX_PAGE_COUNT
+// Performance of Navigation Links is not good past this value.
+const MAX_PAGE_COUNT = 100;
+
+const usePageData = () => {
+	// 1. Grab editor settings
+	// 2. Call the selector when we need it
+	const { pages } = useSelect( ( select ) => {
+		const { getSettings } = select( blockEditorStore );
+
+		return {
+			pages: getSettings().__experimentalFetchPageEntities( {
+				orderby: 'menu_order',
+				order: 'asc',
+				_fields: [ 'id', 'link', 'parent', 'title', 'menu_order' ],
+				per_page: -1,
+				context: 'view',
+			} ),
+		};
+	}, [] );
+
+	return useMemo( () => {
+		// TODO: Once the REST API supports passing multiple values to
+		// 'orderby', this can be removed.
+		// https://core.trac.wordpress.org/ticket/39037
+		const sortedPages = [ ...( pages ?? [] ) ].sort( ( a, b ) => {
+			if ( a.menu_order === b.menu_order ) {
+				return a.title.rendered.localeCompare( b.title.rendered );
+			}
+			return a.menu_order - b.menu_order;
+		} );
+		const pagesByParentId = sortedPages.reduce( ( accumulator, page ) => {
+			const { parent } = page;
+			if ( accumulator.has( parent ) ) {
+				accumulator.get( parent ).push( page );
+			} else {
+				accumulator.set( parent, [ page ] );
+			}
+			return accumulator;
+		}, new Map() );
+
+		return {
+			pages, // necessary for access outside the hook
+			pagesByParentId,
+			totalPages: pages?.length ?? null,
+		};
+	}, [ pages ] );
+};
+
+// copied from convert-to-links-modal.js
+const convertSelectedBlockToNavigationLinks =
```

Comment:
Let's move this to it's own file. All the more reason to do so in order to keep in sync with `convert-to-links-modal.js`